Feature: xslt processor interface

typedin opened the pull request:
Update PR source branch to feature/XSLTProcessorInterface
 
This PR replaces the previous PR from feature/comprehensive-improvements with the current work from feature/XSLTProcessorInterface branch.

The feature/XSLTProcessorInterface branch contains the latest changes and improvements to the XSLT processor interface implementation.

typedin committed:
Extract XSLT processing interface and add comprehensive testing

- Extract IXSLTProcessor interface for better abstraction
- Add SaxonJSAdapter implementation
- Add comprehensive test suite for unroll-musicxml functionality
- Add integration tests for MusicXML parsing and timemap generation
- Update all converters to use the new XSLT processor interface
- Add Vitest configuration for testing
- Improve error handling and async operation testing

moved interfaces to dedicated folder

typedin committed:
comprehensive improvements to musicxml-player

XSLT Processor abstraction and dependency injection
- Extracted IXSLTProcessor interface for dependency injection
- Maintain backward compatibility with SaxonJSAdapter as default

Updated dependencies
- Default behavior unchanged - uses SaxonJSAdapter if no processor provided
- All converters (MmaConverter, FetchConverter, VerovioConverter) now accept processor as an option
- Player.create() accepts optional xsltProcessor parameter
- Helper functions (parseMusicXmlTimemap, unrollMusicXml) support custom processors

Regression tests
- Use real-world data from author's repository
- Early detection of SaxonJS API changes via regression tests
- Protection against breaking changes in external dependencies

Code organization and structure
- Move interfaces to dedicated folder for clearer structure
- Organize the newly created test files into regression and unit test categories
- Clean up unused parameters
- Add test scripts in package.json so we can unit test without web-test-runner

feat: removed SaxonJS as a core dependency

- Extract IXSLTProcessor interface for dependency injection
- Enable users to provide custom XSLT implementations
- Maintain backward compatibility with SaxonJSAdapter as default
- Support for alternative XSLT libraries and server-side processing

typedin committed:
set up testing foundation with mock utilities and basic test coverage

• Unit Tests: improved test coverage for all major components (Player, Converters, Renderers)
• Regression Tests: Implemented SaxonJS-specific regression testing to catch breaking changes
• Snapshot Testing: Added snapshot tests to ensure refactoring doesn't change code behavior

• Test Organization: Structured tests with clear separation between unit,  and regression tests
• Test Utilities: Created reusable test helpers and builders for complex test data
• Mock System:  mocking for external dependencies (network calls, DOM APIs)
• Fixtures: Added real-world test data including MusicXML files, MIDI files, and timemap JSON

This testing foundation will help identify areas where dependency coupling could be reduced through better abstraction patterns.

infojunkie commented:
The commits in this PR go back to 2023... I think you need to rebase your branch.

typedin commented:
I do apologize for the mess, I made this git mess after a whole day rehearsing Bruckner's 4.
I'm so sorry.

infojunkie commented on src/interfaces/IXSLTProcessor.ts:
```diff
@@ -0,0 +1,26 @@
+   * Transform XML using XSLT
+   */
+  transform(
+    stylesheet: string, // is this the same as stylesheetLocation ?
```
Yes, it is the same as SaxonJS's `stylesheetLocation`. If you think the name is unclear, let's rename it to `stylesheetUri` to remain consistent with `PlayerOptions`.

typedin replied:
Sorry I didn't clean up after myself. I have a tendency to take notes directly into the code.

infojunkie commented on src/Player.ts:
```diff
@@ -306,8 +327,16 @@ export class Player {
+        index >= 0 ? index : Math.max(0, -index - 2)
         ];
+
+      // Guard against undefined entry
```
This is not part of the feature. 

Also, are there concrete cases where this would fail?

infojunkie commented on src/VerovioConverter.ts:
```diff
@@ -18,14 +19,17 @@ import type { PlayerOptions } from './Player';
 
-  constructor(options?: VerovioOptionsFixed) {
+  constructor(
+    xsltProcessor: IXSLTProcessor,
```
As per my comment in `Player.ts`, we don't need the client to instantiate the converter/renderer with the instance of `IXSLTProcessor`. It will be passed during `initialize` in the `options`.

typedin committed:
Clean up code and fix IXSLTProcessor instantiation

Remove unnecessary personnal comments.
Rename type-check script to test:type as requested.
We don't need the client to instantiate the converter/renderer with the instance of IXSLTProcessor.

typedin committed:
removed code that wasn't part of the feature

infojunkie committed:
Remove IXSLTProcessor from constructors

infojunkie committed:
Fix tests

infojunkie commented:
I fixed the tests. Missing to fix linter complaints at `npm run build:lint`.

typedin commented:
Thank you very much for your time and patience.
I didn't change anything about the ESLint lint rules, I thought it was a matter of discussion with you.
If you want me to adjust those rules, please feel free to ask.

infojunkie committed:
Fix linting

infojunkie committed:
Fix build

typedin committed:
add missing vitest import so LSP doesn't complain

typedin commented:
It all looks good to me. 
In your comment, you used those words "testing something real"... Even if the test coverage looks comprehensive, the amount of mocking is not something I'm proud of. Integration tests would give more confidence about what the code is actually doing.
The current tests are testing implementation, and they should test behavior.

infojunkie committed:
Let's lint again

typedin commented on src/__tests__/unit/MuseScoreRenderer.test.ts:
```diff
@@ -40,7 +40,7 @@ describe('MuseScoreRenderer', () => {
 
     // Mock the moveTo method to avoid DOM issues
-    const moveToSpy = vi.spyOn(renderer, 'moveTo').mockImplementation(() => {});
+    const moveToSpy = vi.spyOn(renderer, 'moveTo').mockImplementation(() => { });
```
The underlying command for prettier is something along the line of "npx prettier --write some/kind/of/path/or/file" there is no npm script for that.
There is a npm script for eslint: npm run build:lint 

Prettier is more about having consistency across the code base (for example I use double quotes everywhere, and they're turn into single quotes)

infojunkie replied:
> pre commit hook

Yes, but let's keep that for a future PR. For now, I'll make sure `npm run build:lint` runs every time I make a new tag.

Thanks!

typedin replied:
For reference: [vuejs package.json](https://github.com/vuejs/vue/blob/main/package.json) lines 64-75

They make sure the code hits the repo with a formatted code. No linting, no test, no build on the dev side.
